swarmrl/observables/subdivided_vision_cones.py

```diff
@@ -0,0 +1,221 @@
+    """
+    The vision cone acts like a camera for the particles. It can either output all
+    colloids within its view, a function of the distances between one colloid and all
+    other colloids, or the normalised distance to the source.
+    """
+
+    def __init__(
+        self,
+        vision_range: float,
+        vision_half_angle: float,
+        n_cones: int,
+        radii: List[float],
+        detected_types=None,
+    ):
+        """
+        Constructor for the observable.
+
+        Parameters
+        ----------
+        vision_range : float
+                How far the particles can see.
+        vision_half_angle : float
+                the half width of the field of view in radiant units
+        n_cones: int
+                In how many cone is the field of view subdivided
+        radii: list
+                List of the radii of the colloids in the experiment,
+                ordered by the ids of the colloids
+        detected_types: list
+                list of colloid types to be detected in requested order.
+                For example [0,2] here colloids of type 1 won't be detected.
+                if None then al will be detected.
+
+        """
+        self.vision_range = vision_range
+        self.vision_half_angle = vision_half_angle
+        self.n_cones = n_cones
+        self.radii = radii
+        self.detected_types = detected_types
+
+    def _detect_all_things_to_see(self, colloids: List[Colloid]):
+        """
+        Determines what types of colloids are present in the simulation
+
+        Parameters
+        ----------
+        colloids : object
+                Colloids with all possible different types.
+
+        Returns
+        One dimensional np.array with all possible types
+        in the corresponding index in which
+        they will also later be found when the observable is calculated.
+        """
+        all_types = []
+        for c in colloids:
+            if c.type not in all_types:
+                all_types.append(c.type)
+        self.detected_types = np.array(np.sort(all_types))
+
+    def _calculate_director(self, colloid):
+        """
+        Calculates the normalised director of the colloids.
+
+        Parameters
+        ----------
+        colloid : object
+                Colloid for which the observable should be computed.
+
+        Returns
+        Colloid position and vision cone director.
+        """
+        my_pos = np.copy(colloid.pos)
+        my_director = np.copy(colloid.director)
+        # TODO: Add 3D support
+        my_director = my_director / np.linalg.norm(my_director)
+        assert abs(colloid.pos[2]) < 10e-6
+        assert abs(colloid.director[2]) < 10e-6
+        return my_pos, my_director
+
+    def _calculate_cones_single_object(
+        self, my_pos, my_director, c_type, c_pos, radius
+    ):
+        # generate a blue print of the output values
+        vision_val_out = jnp.ones((self.n_cones, len(self.detected_types)))
```

Comment:
If you used `np.zeros` here there would be no memory usage until the array was populated. This is a truly unnecessary thing or us to do I just wanted to point it out.